core/templates/services/html-length.service.ts

```diff
@@ -50,51 +65,129 @@ export class HtmlLengthService {
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    /* CustomTagsLength is being passed to calculateBaselineLength()
+    method to ensure try-catch block should not raise error if
+    we use non-interactive RTE */
+    let totalWords = this.calculateBaselineLength(
+      sanitizedHtml, CALCULATION_TYPE_WORD, customTagsLength);
+
+    if (customTags) {
+      for (const customTag of customTags) {
+        totalWords += this.getWeightForNonTextNodes(
+          customTag, CALCULATION_TYPE_WORD);
+      }
+    }
+    return totalWords;
+  }
+
+  computeHtmlLengthInCharacters(htmlString: string): number {
+    if (!htmlString) {
+      this.loggerService.error('Empty string was passed to compute length');
+      return 0;
+    }
 
+    const sanitizedHtml = this.sanitizer.sanitize(
+      SecurityContext.HTML, htmlString) as string;
     // Identify custom tags using regex on the original HTML string.
-    const customTagRegex = /<oppia-noninteractive-(?:math|image)[^>]*>/g;
-    const customTags = htmlString.match(customTagRegex);
+    const customTags = htmlString.match(CUSTOM_TAG_REGEX);
+    let customTagsLength = (customTags) ? customTags.length : 0;
+
+    /* CustomTagsLength is being passed to calculateBaselineLength()
+    method to ensure try-catch block should not raise error if
+    we use non-interactive RTE */
+    let totalCharacters = this.calculateBaselineLength(
+      sanitizedHtml, CALCULATION_TYPE_CHARACTER, customTagsLength);
 
     if (customTags) {
       for (const customTag of customTags) {
-        totalWeight += this.getWeightForNonTextNodes(customTag);
+        totalCharacters += this.getWeightForNonTextNodes(
+          customTag, CALCULATION_TYPE_CHARACTER);
       }
     }
-    return totalWeight;
+    return totalCharacters;
   }
 
-  private calculateBaselineLength(sanitizedHtml: string): number {
+  calculateBaselineLength(
+      sanitizedHtml: string,
+      calculationType: CalculationType, CustomTagsLength: number): number {
     let domparser = new DOMParser();
-    let dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+    let dom: Document;
+    try {
+      dom = domparser.parseFromString(sanitizedHtml, 'text/html');
+      if (dom.body.children.length === 0 && !CustomTagsLength) {
+        throw new Error(
+          'No HTML tags found. Ensure ' +
+          'that a valid string that includes HTML tags is provided.');
+      }
+    } catch (error) {
+      throw new Error(
+        'Failed to parse HTML string.' +
+        ' Ensure that a valid string that includes HTML tags is provided.');
+    }
     let totalWeight = 0;
     for (let tag of Array.from(dom.body.children)) {
       const ltag = tag.tagName.toLowerCase();
       if (this.textTags.includes(ltag)) {
-        totalWeight += this.getWeightForTextNodes(tag as HTMLElement);
+        totalWeight +=
+        this.getWeightForTextNodes(tag as HTMLElement, calculationType);
```

Comment:
Please fix indentation.

```suggestion
        totalWeight += this.getWeightForTextNodes(
          tag as HTMLElement, calculationType);
```

Reply:
Done!